### PR TITLE
Support hitCount breakpoint

### PR DESCRIPTION
**Pull request by testforstephen**
Signed-off-by: Jinbo Wang <jinbwan@microsoft.com>

### CODE REVIEW

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/BreakpointManager.java [hunk 1]
```diff
@@ -174,4 +178,11 @@ public void reset() {
         this.breakpoints.clear();
         this.nextBreakpointId.set(1);
     }
+
+    private static String formatPath(String source) {
+        if (AdapterUtils.isWindows()) {
```
Comment (akaroml) [inline org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/BreakpointManager.java]: Windows is not sensitive to cases in file paths. Why do we still have to convert everything to lower case?

**Review by akaroml — CHANGES_REQUESTED** [review]

**Review by akaroml — CHANGES_REQUESTED** [review]
Please remove the logic to handle OS specific path issues. These should be handled by VSCode. At the same time, we can use URI as breakpoint key to resolve the issue at our side.

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java [hunk 2]
```diff
@@ -53,7 +55,7 @@
     private Consumer<Events.DebugEvent> eventConsumer;
 
     private boolean debuggerLinesStartAt1 = true;
-    private boolean debuggerPathsAreURI = false;
+    private boolean debuggerPathsAreURI = true;
```
Comment (akaroml) [inline org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java]: debuggerPathsAreURI => debuggerPathsAreUri

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java [hunk 3]
```diff
@@ -312,9 +314,27 @@ public DebugAdapter(Consumer<Events.DebugEvent> consumer, IProviderContext conte
     }
 
     private Responses.ResponseBody setBreakpoints(Requests.SetBreakpointArguments arguments) {
+        String clientPath = arguments.source.path;
+        if (AdapterUtils.isWindows() && clientPath.length() > 2) {
+            // VSCode may send drive letters with inconsistent casing which will mess up the key
+            // in the BreakpointManager. See https://github.com/Microsoft/vscode/issues/6268
+            // Normalize the drive letter casing. Note that drive letters
+            // are not localized so invariant is safe here.
+            String drive = clientPath.substring(0, 2);
```
Comment (akaroml) [inline org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java]: This does not look good. Please consider org.apache.commons.io.FilenameUtils.getPrefix or use regular expression.

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java [hunk 4]
```diff
@@ -53,7 +55,7 @@
     private Consumer<Events.DebugEvent> eventConsumer;
 
     private boolean debuggerLinesStartAt1 = true;
-    private boolean debuggerPathsAreURI = false;
+    private boolean debuggerPathsAreURI = true;
     private boolean clientLinesStartAt1 = true;
     private boolean clientPathsAreURI = false;
```
Comment (akaroml) [inline org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java]: clientPathsAreURI => clientPathsAreUri

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java (hunk 3 above)
Comment (akaroml) [reply org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java]: We use URIs internally. So the drive letters of different cases won't result in duplicated breakpoints. Do we still need to keep normalizing the drive letter?

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java [hunk 5]
```diff
@@ -616,25 +698,45 @@ private int convertClientLineToDebugger(int line) {
...
-                sourcepath = Paths.get(this.cwd, originalSourcePath).toString();
+        String sourceName = "";
+        String relativeSourcePath = "";
+        try {
+            // When the .class file doesn't contain source information in meta data,
+            // invoking Location#sourceName() would throw AbsentInformation.
+            sourceName = location.sourceName();
+            relativeSourcePath = location.sourcePath();
+        } catch (AbsentInformationException e) {
+            int dollar = fullyQualifiedName.indexOf('$');
```
Comment (akaroml) [inline org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java]: Please use a regular expression for matching strings.

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/jdt/JdtSourceLookUpProvider.java [hunk 6]
```diff
@@ -63,22 +62,27 @@ public void initialize(Map<String, Object> props) {
...
+                    IJavaElement javaElement = typeRoot.getElementAt(offset);
+                    if (javaElement instanceof SourceField || javaElement instanceof SourceMethod
+                            || javaElement instanceof BinaryMember) {
+                        IType type = ((IMember) javaElement).getDeclaringType();
+                        fqn = type.getFullyQualifiedName();
+                    } else if (javaElement instanceof SourceType) {
+                        fqn = ((SourceType) javaElement).getFullyQualifiedName();
+                    }
+                } catch (JavaModelException e) {
+                    Logger.logException("Add breakpoint exception", e);
```
Comment (akaroml) [inline org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/jdt/JdtSourceLookUpProvider.java]: Why should the getFullyQualifiedName utility function throw breakpoint exception?

**Review by akaroml — CHANGES_REQUESTED** [review]

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/jdt/JdtSourceLookUpProvider.java (hunk 6 above)
Comment (testforstephen) [reply org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/jdt/JdtSourceLookUpProvider.java]: `IJavaElement javaElement = typeRoot.getElementAt(offset);` may throw JavaModelException. Current policy is to eat the exception. That may be not the correct handing policy. I'll choose to throw it out and let dispatcher send an error response to vscode.

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java (hunk 5 above)
Comment (testforstephen) [reply org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java]: agree.

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java (hunk 2 above)
Comment (testforstephen) [reply org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java]: fixed.

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java (hunk 4 above)
Comment (testforstephen) [reply org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java]: fixed

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java (hunk 3 above)
Comment (testforstephen) [reply org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java]: fixed.

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java [hunk 7]
```diff
@@ -315,14 +318,16 @@ public DebugAdapter(Consumer<Events.DebugEvent> consumer, IProviderContext conte
...
-        if (AdapterUtils.isWindows() && clientPath.length() > 2) {
+        if (AdapterUtils.isWindows()) {
             // VSCode may send drive letters with inconsistent casing which will mess up the key
             // in the BreakpointManager. See https://github.com/Microsoft/vscode/issues/6268
             // Normalize the drive letter casing. Note that drive letters
             // are not localized so invariant is safe here.
-            String drive = clientPath.substring(0, 2);
-            if (Character.isLowerCase(drive.charAt(0)) && drive.charAt(1) == ':') {
-                clientPath = clientPath.replaceFirst(drive, drive.toUpperCase());
+            String drivePrefix = FilenameUtils.getPrefix(clientPath);
```
Comment (akaroml) [inline org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java]: Here's the case of one day VSCode supports URI and then pass the path in URI format. e.g. file:///c:\abc.java.
Then the prefix you get is actually file:/// but not c:\.
That's why I suggest a second check on the prefix to use regex to match for drive info.

**Review by akaroml — CHANGES_REQUESTED** [review]

**Review by akaroml — APPROVED** [review]

**Pull request merged**